build/xtask/src/sizes.rs

```diff
@@ -79,233 +68,235 @@ pub fn run(
             }
+            if !printed_name {
+                printed_name = true;
+                writeln!(out, "{}:", name)?;
+            }
+            write!(out, "  {:<6} {: >5}", format!("{}:", mem), suggestion)?;
+            out.set_color(ColorSpec::new().set_dimmed(true))?;
+            if name == "kernel" {
+                writeln!(out, " (currently {})", size)?;
+                *savings.entry(mem).or_default() += size as u64 - suggestion;
+            } else {
+                writeln!(out, " (autosized with max {})", size)?;
+            }
+            out.reset()?;
         }
+    }
 
-        if !only_suggest {
-            out.set_color(ColorSpec::new().set_bold(true))?;
-            writeln!(out, "\n------------ Total savings ------------")?;
-            out.reset()?;
-            for (mem, savings) in &savings {
-                writeln!(out, "  {:<6} {}", format!("{}:", mem), savings)?;
-            }
+    if !only_suggest && !savings.is_empty() {
+        out.set_color(ColorSpec::new().set_bold(true))?;
+        writeln!(out, "\n------------ Total savings ------------")?;
+        out.reset()?;
+        for (mem, savings) in &savings {
+            writeln!(out, "  {:<6} {}", format!("{}:", mem), savings)?;
         }
     }
 
     Ok(())
 }
 
-fn create_sizes(cfg: &Path) -> anyhow::Result<TaskSizes> {
-    let toml = Config::from_file(&cfg)?;
-
-    let dist_dir = Path::new("target").join(&toml.name).join("dist");
-
-    let mut memories = IndexMap::new();
-    for (name, out) in &toml.outputs {
-        // This is called after `dist`, which logs this overflow gracefully
-        let end = out.address.checked_add(out.size).unwrap();
-        memories.insert(name.clone(), out.address..end);
-    }
-
-    let output_region = |vaddr: u64| {
-        memories
-            .iter()
-            .find(|(_, region)| region.contains(&vaddr.try_into().unwrap()))
-            .map(|(name, _)| name.as_str())
-    };
-
-    let suggest = if toml.target.starts_with("thumbv7")
-        || toml.target.starts_with("thumbv6m")
-    {
-        pow2_suggest
-    } else if toml.target.starts_with("thumbv8m") {
-        armv8m_suggest
-    } else {
-        panic!("Unknown target: {}", toml.target);
+/// Loads the size of the given task (or kernel)
+pub fn load_task_size<'a>(
+    toml: &'a Config,
+    name: &str,
+    stacksize: u32,
+) -> Result<IndexMap<&'a str, u64>> {
+    // Load the .tmp file (which does not have flash fill) for everything
+    // except the kernel
+    let elf_name =
+        Path::new("target")
+            .join(&toml.name)
+            .join("dist")
+            .join(match name {
+                "kernel" => name.to_owned(),
+                _ => format!("{}.tmp", name),
+            });
+
+    let buffer = std::fs::read(elf_name)?;
+    let elf = match Object::parse(&buffer)? {
+        Object::Elf(elf) => elf,
+        o => bail!("Invalid Object {:?}", o),
     };
 
-    let check_task = move |name: &str,
```

Comment:
i am glad this monstrosity is leaving, haha. i *almost* wanted to turn it into a free function, but it captured one or two variables and that was convenient.